### PR TITLE
WIP: Function factory

### PR DESCRIPTION
**Pull request by vincentarelbundock**
This is a first draft for two new features:

1. `custom_dict` can take two attributes: `origin_regex` and `origin_valid`. Both must be either `NULL` (in which case they are ignored) or character vectors. When `origin` is found in `attr(custom_dict, "origin_regex")`, the `origin_regex` is flipped `TRUE` automatically. When `origin` is *not* found in `attr(custom_dict, "origin_valid")`, we exit with an error.
2. `countrycode_factory` is a function factory that allows users to create new functions with different default arguments, or with a default dictionary.

Usage examples are given in the README of the `functionfactory` branch:

https://github.com/vincentarelbundock/countrycode/tree/functionfactory#custom-dictionary-attributes

This still needs documentation, etc., but I'd be very eager to get feedback on this, if anyone has time to take it for a spin. @cjyetman , maybe?

### CODE REVIEW

File: R/countrycode_factory.R [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  # default function
+  out = countrycode
+
+  # change defaults
+  if(!is.null(origin)) formals(out)$origin <- origin
+  if(!is.null(destination)) formals(out)$destination <- destination
+  if(!is.null(warn)) formals(out)$warn <- warn
+  if(!is.null(nomatch)) formals(out)$nomatch <- nomatch
+  if(!is.null(custom_dict)) formals(out)$custom_dict <- custom_dict
+  if (!is.null(custom_match)) formals(out)$custom_match <- custom_match
```
Comment (salim-b) [inline R/countrycode_factory.R:22]: You could make this generic:

```r
args <- formals()

for (name in names(args)) {
  if(!is.null(args[[name]])) formals(out)[[name]] <- args[[name]])
}
```

Comment (vincentarelbundock) [reply R/countrycode_factory.R]: Good idea, thanks!

Comment (vincentarelbundock) [reply R/countrycode_factory.R]: turns out that tests fail when I do this :(  Will try to think of another more generic approach eventually

Comment (salim-b) [reply R/countrycode_factory.R:61]: Sorry, my fault! I should have tested the proposed code 🙈

I realized that `formals()` doesn't return the actual arguments but the _formal_ ones of the function definition.

This actually works:

```r
for (name in names(formals())) {
    arg <- get(name)
    if (!is.null(arg)) {
        formals(out)[[name]] <- arg
    }
}
```

Comment (vincentarelbundock) [reply R/countrycode_factory.R:61]: looks like this works, thanks!

**Pull request merged**